pkg/ebpf/flow_filter.go

```diff
@@ -0,0 +1,201 @@
+package ebpf
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/cilium/ebpf"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+type FlowFilterConfig struct {
+	FlowFilterDirection       string
+	FlowFilterIPCIDR          string
+	FlowFilterProtocol        string
+	FlowFilterSourcePort      intstr.IntOrString
+	FlowFilterDestinationPort intstr.IntOrString
+	FlowFilterPort            intstr.IntOrString
+	FlowFilterIcmpType        int
+	FlowFilterIcmpCode        int
+	FlowFilterPeerIP          string
+	FlowFilterAction          string
+}
+
+type FlowFilter struct {
+	// eBPF objs to create/update eBPF maps
+	objects *BpfObjects
+	config  *FlowFilterConfig
+}
+
+func NewFlowFilter(objects *BpfObjects, cfg *FlowFilterConfig) *FlowFilter {
+	return &FlowFilter{
+		objects: objects,
+		config:  cfg,
+	}
+}
+
+func (f *FlowFilter) ProgramFlowFilter() error {
+	log.Infof("Flow filter config: %v", f.config)
+	key, err := f.getFlowFilterKey(f.config)
+	if err != nil {
+		return fmt.Errorf("failed to get flow filter key: %w", err)
+	}
+
+	val, err := f.getFlowFilterValue(f.config)
+	if err != nil {
+		return fmt.Errorf("failed to get flow filter value: %w", err)
+	}
+
+	err = f.objects.FilterMap.Update(key, val, ebpf.UpdateAny)
+	if err != nil {
+		return fmt.Errorf("failed to update flow filter map: %w", err)
+	}
+
+	log.Infof("Programmed flow filter with key: %v, value: %v", key, val)
+
+	return nil
+}
+
+func (f *FlowFilter) getFlowFilterKey(config *FlowFilterConfig) (BpfFilterKeyT, error) {
+	key := BpfFilterKeyT{}
+
+	ip, ipNet, err := net.ParseCIDR(config.FlowFilterIPCIDR)
+	if err != nil {
+		return key, fmt.Errorf("failed to parse FlowFilterIPCIDR: %w", err)
+	}
+	if ip.To4() != nil {
+		copy(key.IpData[:], ip.To4())
+	} else {
+		copy(key.IpData[:], ip.To16())
+	}
+	pfLen, _ := ipNet.Mask.Size()
+	key.PrefixLen = uint32(pfLen)
+
+	return key, nil
+}
+
+func (f *FlowFilter) getFlowFilterValue(config *FlowFilterConfig) (BpfFilterValueT, error) {
+	val := BpfFilterValueT{}
+
+	switch config.FlowFilterDirection {
+	case "Ingress":
+		val.Direction = BpfDirectionTINGRESS
+	case "Egress":
+		val.Direction = BpfDirectionTEGRESS
+	default:
+		val.Direction = BpfDirectionTMAX_DIRECTION
+	}
+
+	switch config.FlowFilterAction {
+	case "Reject":
+		val.Action = BpfFilterActionTREJECT
+	case "Accept":
+		val.Action = BpfFilterActionTACCEPT
+	default:
+		val.Action = BpfFilterActionTMAX_FILTER_ACTIONS
```

Comment:
since this is a required parameter, shouldn't we fail here and avoid having to deal with invalid config in the bpf program?

Reply:
the ebpf code has handle for this special value to just ignore filtering all together plus if we provision from the operator taht won't happen